src/socket.c

```diff
@@ -696,4 +696,172 @@ void pthreads_socket_free(pthreads_socket_t *socket, zend_bool closing) {
 
 	efree(socket);
 }
+
+void pthreads_socket_recvfrom(zval *object, zval *buffer, zend_long len, zend_long flags, zval *name, zval *port, zval *return_value) {
+	pthreads_object_t *threaded =
+		PTHREADS_FETCH_FROM(Z_OBJ_P(object));
+
+	socklen_t			slen;
+	int					retval;
+	zend_string *recv_buf;
+
+	recv_buf = zend_string_alloc(len + 1, 0);
+
+	switch (threaded->store.sock->type) {
+#ifndef _WIN32
+		case AF_UNIX: {
+			struct sockaddr_un	s_un;
+
+			slen = sizeof(s_un);
+			s_un.sun_family = AF_UNIX;
+			retval = recvfrom(threaded->store.sock->fd, ZSTR_VAL(recv_buf), len, flags, (struct sockaddr *)&s_un, (socklen_t *)&slen);
+
+			if (retval < 0) {
+				zend_string_free(recv_buf);
+				PTHREADS_SOCKET_ERROR();
+			}
+			ZSTR_LEN(recv_buf) = retval;
+			ZSTR_VAL(recv_buf)[ZSTR_LEN(recv_buf)] = '\0';
+
+			zval_dtor(buffer);
+			zval_dtor(name);
+
+			ZVAL_NEW_STR(buffer, recv_buf);
+			ZVAL_STRING(name, s_un.sun_path);
+		} break;
+#endif
+		case AF_INET: {
+			struct sockaddr_in	sin;
+			char				*address;
+
+			slen = sizeof(sin);
+			memset(&sin, 0, slen);
+			sin.sin_family = AF_INET;
+
+			if (port == NULL) {
+				zend_string_free(recv_buf);
+				WRONG_PARAM_COUNT;
+			}
+
+			retval = recvfrom(threaded->store.sock->fd, ZSTR_VAL(recv_buf), len, flags, (struct sockaddr *)&sin, (socklen_t *)&slen);
+
+			if (retval < 0) {
+				zend_string_free(recv_buf);
+				PTHREADS_SOCKET_ERROR();
+			}
+			ZSTR_LEN(recv_buf) = retval;
+			ZSTR_VAL(recv_buf)[ZSTR_LEN(recv_buf)] = '\0';
+
+			zval_dtor(buffer);
+			zval_dtor(name);
+			zval_dtor(port);
+
+			address = inet_ntoa(sin.sin_addr);
+
+			ZVAL_NEW_STR(buffer, recv_buf);
+			ZVAL_STRING(name, address ? address : "0.0.0.0");
+			ZVAL_LONG(port, ntohs(sin.sin_port));
+		} break;
+#if HAVE_IPV6
+		case AF_INET6: {
+			struct sockaddr_in6	sin6;
+			char				addr6[INET6_ADDRSTRLEN];
+
+			slen = sizeof(sin6);
+			memset(&sin6, 0, slen);
+			sin6.sin6_family = AF_INET6;
+
+			if (port == NULL) {
+				efree(recv_buf);
```

Comment:
shouldn't this be `zend_string_free()`? (it is also the same in PHP sockets though)

Reply:
Yes, you're right. This should be ```zend_string_free()``, changed.